### PR TITLE
feat(cabling): end-to-end Cable + GenericObject termination ingest (OBS-3370, OBS-1080)

### PR DESCRIPTION
**Pull request by leoparente**
## Summary

End-to-end **cabling ingest** for Diode (OBS-3370), and the fix for the OBS-1080 crash (`TypeError: unhashable type: 'dict'` / HTTP 500 when ingesting a `Cable` with `a_terminations`/`b_terminations`). Cables now flow through the full plan → apply pipeline: a `Cable` with `GenericObject` terminations creates/updates the correct `dcim_cable` + `dcim_cabletermination` rows, idempotently and A/B-order-insensitive.

## What changes

- **Transformer** — extracts each `GenericObject` termination to `{object_type, object_id: UnresolvedReference}`, dispatching on the variant key via `get_generic_object_variant` (nested variant keys are snake-case-normalized, so camelCase protoJSON like `objectInterface` resolves too); records termination uuids in `_refs` (terminated objects are created before the cable); unknown variant keys and known-but-unsupported object types warn (with distinct messages) + skip that single termination (never crash, never drop the cable).
- **Matcher** — new `CableTerminationSetMatcher`: matches a cable by its **canonical, A/B-insensitive termination set** (`build_queryset` does an exact `CableTermination`-set match, rejecting subset/superset — including multi-termination-per-end). `dcim.cable` added to `_REQUIRES_PRE_SAVE_MATCH`; `_find_obj_cache_key` returns `None` for cable (its identity lives in list fields the scalar cache key skips).
- **Applier** — resolves termination refs to `{object_type, object_id: pk}` (numeric list-index path support in `_get_path`/`_set_path`; deep-copy of change data) and saves via the existing `CableSerializer`, whose `save() → update_terminations()` auto-creates the `CableTermination` rows (no separate apply). `Cable.clean()` conflicts (already-cabled / both-ends-required / incompatible types) surface as clean per-entity deviations. A dangling `new_object:` reference at apply time raises a clean, key-named per-entity error rather than a 500; unrelated `KeyError`s propagate as real bugs.
- **Differ** — captures the property-backed `a_terminations`/`b_terminations` with stable ordering so re-diff of an applied cable is a no-op (idempotency). Because cable identity is A/B-insensitive, a re-feed with the ends swapped is realigned to the existing assignment (`_align_cable_ends`) and converges as a NOOP instead of toggling `cable_end` forever.

## Certified scope

All 9 single-object termination types: interface, front port, rear port, console port, console-server port, power port, power outlet, power feed, circuit termination. Identity = A/B-insensitive termination set; `type`/`status`/`label`/`length`/`tenant` are mutable attributes (update, never fork). Conflicts fail the change as a per-entity deviation.

## Validation

- Full v4.6.x plugin suite: **419 tests OK** (transformer, matcher DB-backed set-match incl. multi-termination-per-end, applier end-to-end apply, integration fixtures across all termination types).
- **Live e2e (validated during development, harness not included in this PR)**: with a local real-OAuth + HTTP + NetBox harness, two A/B-swapped `bulk-plan`+`bulk-apply` requests converged to **one** cable, and an already-cabled interface returned a clean HTTP 207 deviation (no 500). The shipped regression coverage for these behaviors lives in the v4.6.x suite above (matcher set-match + applier end-to-end apply + integration idempotency).
- **Review hardening** — two adversarial review passes plus several Codex review rounds. Blocking edge-case bugs found and fixed (nested-termination-ref dedup, re-sorted-termination int at ref path, an order-fragile DB test), each with regression coverage; review follow-ups addressed: camelCase nested variant keys, A/B end-swap NOOP convergence, narrowed apply-time `KeyError` handling with actionable messages, `fingerprint` return-type annotation, distinct skip warnings, and removal of a `noqa: C901` by extracting a helper. Findings judged invalid were refuted with tests (e.g. multi-termination `build_queryset` matching; NetBox's `Cable(a_terminations=…, b_terminations=…)` constructor).

## Performance (hot-path audit)

**No material degradation for non-cable entities.** The plugin hot path (`transform → diff → apply`, run per-entity for many entities per bulk request) was audited change-by-change against `develop`.

- **All real cabling work is gated** behind `a_/b_terminations` key presence, `object_type == "dcim.cable"`, or membership in a cable-only dict — so common entities (devices, interfaces, IPs, prefixes) never enter the set-match query, termination sort, or generic-object variant paths. Their only added cost is a handful of O(1) probes per node/key (tuple-membership `==`, `dict.get`→`None`, a frozenset miss), each sub-microsecond and dwarfed by the surrounding `harmonize_formats` / `find_existing_object` / serializer work already in those loops.
- **The one unconditional per-entity change** is `_pre_apply` switching from a shallow `.copy()` to `copy.deepcopy(change.data)` on the apply path. Benchmarked at ~10.7 µs (typical payload) to ~41 µs (large) per call — i.e. **~1–4 ms per 100 entities, ~10–41 ms per 1000**. Each call is immediately followed by `serializer.is_valid()` + `save()` (a DB write, hundreds of µs to ms), so the deepcopy is **&lt;1–4 % of per-change cost** and, scaling per-change, never becomes the bottleneck. It is also **correctness-required**: ref resolution mutates nested containers (e.g. `a_terminations.0.object_id`) in place, and a shallow copy would corrupt `change.data` on retry (the data is plain JSON — no model instances — so deepcopy carries no hidden expense).

**Conclusion: no change warranted on performance grounds.** (Optional future polish: deepcopy only when a `new_refs` path is multi-segment/nested — cable terminations — and keep the shallow copy for single-segment FK refs; the measured impact doesn't justify it today.)

## Known limitations / backlog

- **Multi-object-per-end (LAG/breakout) UPDATE**: ~~out of certified scope; fails cleanly~~ **resolved** — termination lists are sorted before `new_refs` index paths are computed, so adding a new termination to an existing (netbox_id-matched) cable's end resolves and applies. Regression-tested through the real `generate_changeset → apply_changeset` path.
- **Breakout / profiled-cable connector semantics** (NetBox 4.6 cable profiles / per-termination `connector`) remain out of scope — that needs a proto change to carry per-termination lane info. The termination-set sort is A/B/order-insensitive by design, so it does not preserve the positional lane mapping a profiled cable would need. Tracked as **OBS-3411** (raised in review).

Refs: OBS-3370, OBS-1080.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by leoparente** [comment]
@codex review

File: netbox_diode_plugin/api/applier.py [hunk 1]
```diff
@@ -52,6 +53,13 @@ def apply_changeset(change_set: ChangeSet, request) -> ChangeSetResult:
         except IntegrityError as e:
             logger.error(f"Integrity error {object_type}: {e} {data}")
             raise _err(f"created a conflict with an existing {object_type}", object_type, "__all__")
+        except KeyError as e:
+            # A new_refs path referenced an object that was never created
+            # (e.g. a dangling or rewritten termination reference). Surface a
+            # clean per-entity deviation instead of an uncaught 500 -- the
+            # OBS-1080 failure class this feature set out to eliminate.
+            logger.error(f"unresolved reference applying {object_type}: {e}")
+            raise _err(f"unresolved reference applying {object_type}", object_type, "__all__")
```
[the remark this answers is not in the document]
Comment (leoparente) [reply netbox_diode_plugin/api/applier.py]: Valid — fixed in 1a9a587. The handler now only intercepts a KeyError whose key is a missing `new_object:...` reference lookup (the intended dangling-ref case) and re-raises anything else as a real bug. Added both regression tests: a dangling ref surfaces as a clean `ChangeSetException`, while a KeyError with a non-reference key propagates.

File: netbox_diode_plugin/api/matcher.py [hunk 2]
```diff
@@ -688,6 +695,121 @@ def build_queryset(self, data: dict) -> models.QuerySet:
...
+        reduced = []
+        for field in (self.a_field, self.b_field):
+            for term in data.get(field, []):
+                r = self._reduce(term)
+                if r is None:
+                    return None
+                reduced.append(r)
+        return reduced
+
+    def fingerprint(self, data: dict) -> str | None:
```
[the remark this answers is not in the document]
Comment (leoparente) [reply netbox_diode_plugin/api/matcher.py]: Valid — fixed in 1a9a587: `CableTerminationSetMatcher.fingerprint` is now annotated `int | None` (it returns `hash(...)`). Note the pre-existing matchers in this file (`ObjectMatchCriteria`, `GlobalIPNetworkIPMatcher`, etc.) carry the same `str|None` annotation while returning ints — left untouched as they're develop code outside this PR's scope; happy to fix repo-wide in a follow-up.

File: netbox_diode_plugin/api/transformer.py [hunk 3]
```diff
@@ -199,14 +202,39 @@ def is_supported(field_name, ref_info):
...
+                if ref_info.is_generic_object:
+                    # Each item is a GenericObject dict, e.g. {"object_interface": {...}}.
+                    # Resolve the concrete object_type from the single variant key.
+                    if not isinstance(item, dict) or len(item) != 1:
+                        node['_warnings'][field_name] = node['_warnings'].get(field_name, []) + [
+                            f"Skipping malformed generic-object item (expected single-key dict): {item!r}"
+                        ]
+                        continue
+                    variant_key = next(iter(item))
+                    concrete_type = get_generic_object_variant(variant_key)
```
[the remark this answers is not in the document]
Comment (leoparente) [reply netbox_diode_plugin/api/transformer.py]: Valid — fixed in 1a9a587. `_ensure_snake_case` only normalizes top-level keys, so a camelCase variant key like `objectInterface` was skipped as unknown. The generic-object arm now normalizes the nested variant key with the module's `_camel_to_snake_case` before the variant-map lookup (payload access still uses the raw key). Regression test added: an `aTerminations`/`objectInterface` payload extracts both terminations with no warnings.

File: netbox_diode_plugin/api/differ.py [hunk 4]
```diff
@@ -283,6 +310,15 @@ def _partially_merge(prechange_data: dict, postchange_data: dict, instance) -> d
...
             result[key] = _merge_reference_list(prechange_data.get(key, []), value)
+        elif key in ("a_terminations", "b_terminations") and isinstance(value, list):
+            # Cable termination lists carry no meaningful order (they are a set
+            # of endpoints, not a sequence), but shallow_compare_dict does a
+            # positional list comparison. prechange_data_from_instance() sorts
+            # its side by (object_type, object_id); sort postchange the same
+            # way here so idempotent re-diffs don't spuriously report a change
+            # when the DB's reverse `terminations` relation and the submitted
+            # payload happen to order multi-termination endpoints differently.
+            result[key] = _sorted_termination_refs(value)
```
[the remark this answers is not in the document]
Comment (leoparente) [reply netbox_diode_plugin/api/differ.py:308]: Valid — fixed in 1a9a587. Added `_align_cable_ends`, called after `_partially_merge`: when the submitted ends are a pure swap of the existing assignment (post-A == pre-B and post-B == pre-A), the existing A/B assignment is kept, so the re-diff is a NOOP and alternating feeds converge instead of toggling `cable_end`. Only pure swaps are realigned — that equality also guarantees all refs are resolved pks, so no new_refs index paths can be desynced by the swap; any partial-overlap change still diffs as a genuine UPDATE. DB-backed regression test: seeded cable re-ingested with flipped ends yields an empty change set.

**Comment by leoparente** [comment]
@codex review

File: netbox_diode_plugin/api/applier.py [hunk 5]
```diff
@@ -52,6 +53,15 @@ def apply_changeset(change_set: ChangeSet, request) -> ChangeSetResult:
...
             raise _err(f"created a conflict with an existing {object_type}", object_type, "__all__")
+        except KeyError as e:
+            # Only intercept a missing "new_object:..." reference lookup
+            # (a dangling ref that was never created); surface it as a clean
+            # per-entity error. Any other KeyError is a real bug — re-raise.
+            key = e.args[0] if e.args else None
+            if not (isinstance(key, str) and key.startswith("new_object:")):
+                raise
+            logger.error(f"unresolved reference applying {object_type}: {e}")
+            raise _err(f"unresolved reference applying {object_type}", object_type, "__all__")
```
[the remark this answers is not in the document]
Comment (leoparente) [reply netbox_diode_plugin/api/applier.py]: Fixed in a771b53. The deviation message now includes the missing reference key: `unresolved reference {key} applying {object_type}`, so it's actionable without log access.

File: netbox_diode_plugin/api/transformer.py [hunk 6]
```diff
@@ -199,14 +202,42 @@ def is_supported(field_name, ref_info):
...
+                    raw_variant_key = next(iter(item))
+                    # camelCase protoJSON is normalized at the top level only;
+                    # the nested variant key needs its own normalization.
+                    variant_key = _camel_to_snake_case(raw_variant_key)
+                    concrete_type = get_generic_object_variant(variant_key)
+                    if concrete_type is None or concrete_type not in supported_models:
+                        node['_warnings'][field_name] = node['_warnings'].get(field_name, []) + [
+                            f"Skipping unknown generic-object variant key: {variant_key!r}"
+                        ]
+                        continue
```
[the remark this answers is not in the document]
Comment (leoparente) [reply netbox_diode_plugin/api/transformer.py:237]: Fixed in a771b53. The skip warning is now split into two cases: `Skipping unknown generic-object variant key: {key}` (key not in the variant map) vs `Skipping generic-object variant {key}: {object_type} is not supported in this version.` (valid key, mapped type absent from supported_models) — so the valid-key-but-unsupported-type case is distinguishable.

File: netbox_diode_plugin/api/matcher.py [hunk 7]
```diff
@@ -688,6 +695,120 @@ def build_queryset(self, data: dict) -> models.QuerySet:
...
+
+        qs = self.model_class.objects.annotate(
+            _term_count=models.Count("terminations")
+        ).filter(_term_count=len(pairs))
+        for ct_id, pk in pairs:
+            qs = qs.filter(
+                terminations__termination_type_id=ct_id,
+                terminations__termination_id=pk,
+            )
+        return qs.distinct()
```
[the remark this answers is not in the document]
Comment (leoparente) [reply netbox_diode_plugin/api/matcher.py:818]: Won't do — the current implementation is correct. The premise (constraints AND onto one JOIN alias) doesn't hold here: `build_queryset` issues **one `.filter(terminations__...)` call per pair in a loop**, and Django creates a **separate join alias per `.filter()` call** on a multi-valued relation (the documented [spanning-multi-valued-relationships](https://docs.djangoproject.com/en/stable/topics/db/queries/#spanning-multi-value-relationships) behavior). So each requested (type_id, termination_id) is matched by its own `CableTermination` row — not one row satisfying all pairs. The `_term_count == len(pairs)` guard then rejects supersets, and the per-pair joins reject partial overlaps.

Added DB-backed regression tests in a771b53 that would fail if this were a single-alias AND: `test_multi_termination_per_end_exact_match` (a real 2-terminations-on-one-end cable is matched by its exact 3-set) and `test_multi_termination_superset_rejected`. Both pass. (Conditional aggregation would also work but isn't needed and would be more complex.)

File: netbox_diode_plugin/api/differ.py [hunk 8]
```diff
@@ -295,6 +318,59 @@ def _partially_merge(prechange_data: dict, postchange_data: dict, instance) -> d
...
+
+    object_id may be an int (resolved) or a string (still-unresolved
+    `new_object:...` reference after cleanup_unresolved_references), so the
+    sort key coerces it to str to avoid cross-type comparison errors.
+    """
+    return sorted(
+        refs,
+        key=lambda t: (t.get("object_type", ""), str(t.get("object_id", "")))
+        if isinstance(t, dict)
+        else (str(t),),
```
[the remark this answers is not in the document]
Comment (leoparente) [reply netbox_diode_plugin/api/differ.py:370]: Valid concern, but deferred — it needs proto work that doesn't exist yet, so it's out of this PR's certified scope. Filed as **OBS-3411**.

Three reasons it can't be a quick fix here:
1. **No lane signal in the proto.** There is no per-termination `connector`/`position` field today, so a breakout/profiled cable can't carry lane info regardless — positional order is the only implicit signal and can't be round-tripped reliably.
2. **The sort is load-bearing in scope.** NetBox's reverse `terminations` relation has no ordering guarantee, so both prechange and postchange are sorted to keep idempotency (re-diff of an applied cable = NOOP) and to keep `new_refs` index paths aligned with what the applier resolves. Dropping it reintroduces spurious UPDATEs for ordinary multi-termination cables.
3. **Certified scope** is single-object terminations (all 9 types) + order-insensitive multi-object-per-end (LAG). Breakout/profile connector semantics are already listed as out-of-scope in the PR description.

OBS-3411 captures the full fix (proto per-termination connector + preserve order through transform/diff/apply + matcher check).

**Comment by leoparente** [comment]
@codex review

File: netbox_diode_plugin/tests/v4.6.x/tests/test_applier_cable.py [hunk 9]
```diff
@@ -0,0 +1,278 @@
...
+        # that reuses interface A (now already cabled) cannot be matched by
+        # CableTerminationSetMatcher (different termination set: A<->C vs A<->B),
+        # so it falls through to CableSerializer.is_valid()/Cable.clean(), which
+        # rejects it because interface A already has a cable connection.
+        cs1 = ChangeSet(id="cs-cable-a", changes=[self._cable_change(self.iface_a.pk, self.iface_b.pk)])
+        apply_changeset(cs1, request=None)
+
+        cs2 = ChangeSet(id="cs-cable-b", changes=[self._cable_change(self.iface_a.pk, self.iface_c.pk, label="Cable 2")])
+        with self.assertRaises(Exception) as ctx:
+            apply_changeset(cs2, request=None)
```
[the remark this answers is not in the document]
Comment (leoparente) [reply netbox_diode_plugin/tests/v4.6.x/tests/test_applier_cable.py]: Fixed in f98f4fe. Tightened to `assertRaises(ChangeSetException)` — `apply_changeset` converts the `Cable.clean()` validation failure into `ChangeSetException` (via `error_from_validation_error`), and its `__str__` embeds the errors dict so the conflict-message assertion still holds. An unrelated failure (e.g. a DB error) can no longer make the test pass for the wrong reason.

File: netbox_diode_plugin/tests/v4.6.x/tests/test_cable_matcher.py [hunk 10]
```diff
@@ -0,0 +1,349 @@
...
+        site = Site.objects.create(name="S1", slug="s1")
+        mfr = Manufacturer.objects.create(name="M1", slug="m1")
+        dt = DeviceType.objects.create(manufacturer=mfr, model="DT1", slug="dt1")
+        role = DeviceRole.objects.create(name="R1", slug="r1")
+        dev = Device.objects.create(name="D1", site=site, device_type=dt, role=role)
+        cls.if1 = Interface.objects.create(device=dev, name="eth0", type="1000base-t")
+        cls.if2 = Interface.objects.create(device=dev, name="eth1", type="1000base-t")
+        cls.if3 = Interface.objects.create(device=dev, name="eth2", type="1000base-t")
+        cls.cable = Cable(a_terminations=[cls.if1], b_terminations=[cls.if2])
+        cls.cable.save()
```
[the remark this answers is not in the document]
Comment (leoparente) [reply netbox_diode_plugin/tests/v4.6.x/tests/test_cable_matcher.py:218]: Won't do — this pattern is correct and NetBox-supported. NetBox overrides the model constructor specifically to accept these:

```python
# dcim/models/cables.py
def __init__(self, *args, a_terminations=None, b_terminations=None, **kwargs):
```

So `Cable(a_terminations=[...], b_terminations=[...])` is the intended NetBox API (used throughout NetBox's own test suite), not a non-field kwarg that `Model.__init__` would reject. Empirically it does not raise: this exact form runs in the pre-existing `setUpTestData` (`test_cable_matcher.py:217`) before every test in the class, and the full suite is **419 tests OK** — a `TypeError` at construction would error these tests, not pass them.

File: netbox_diode_plugin/tests/v4.6.x/tests/test_cable_matcher.py [hunk 11]
```diff
@@ -0,0 +1,349 @@
...
+        CableTermination row per requested pair), not a single-alias AND that
+        would make multi-termination matching impossible. Uses fresh
+        interfaces: a termination may only belong to one cable.
+        """
+        dev = self.if1.device
+        m1 = Interface.objects.create(device=dev, name="m0", type="1000base-t")
+        m2 = Interface.objects.create(device=dev, name="m1", type="1000base-t")
+        m3 = Interface.objects.create(device=dev, name="m2", type="1000base-t")
+        multi = Cable(a_terminations=[m1, m2], b_terminations=[m3])
+        multi.save()
```
[the remark this answers is not in the document]
Comment (leoparente) [reply netbox_diode_plugin/tests/v4.6.x/tests/test_cable_matcher.py:311]: Won't do — this pattern is correct and NetBox-supported. NetBox overrides the model constructor specifically to accept these:

```python
# dcim/models/cables.py
def __init__(self, *args, a_terminations=None, b_terminations=None, **kwargs):
```

So `Cable(a_terminations=[...], b_terminations=[...])` is the intended NetBox API (used throughout NetBox's own test suite), not a non-field kwarg that `Model.__init__` would reject. Empirically it does not raise: this exact form runs in the pre-existing `setUpTestData` (`test_cable_matcher.py:217`) before every test in the class, and the full suite is **419 tests OK** — a `TypeError` at construction would error these tests, not pass them.

File: netbox_diode_plugin/tests/v4.6.x/tests/test_cable_matcher.py [hunk 12]
```diff
@@ -0,0 +1,349 @@
...
+        self.assertEqual(list(qs.values_list("pk", flat=True)), [multi.pk])
+
+    def test_multi_termination_superset_rejected(self):
+        """A 2-per-end request must NOT match a cable missing one of them."""
+        dev = self.if1.device
+        m1 = Interface.objects.create(device=dev, name="n0", type="1000base-t")
+        m2 = Interface.objects.create(device=dev, name="n1", type="1000base-t")
+        m3 = Interface.objects.create(device=dev, name="n2", type="1000base-t")
+        # cable only has {m1, m2}; request adds a third (m3)
+        Cable(a_terminations=[m1], b_terminations=[m2]).save()
```
[the remark this answers is not in the document]
Comment (leoparente) [reply netbox_diode_plugin/tests/v4.6.x/tests/test_cable_matcher.py:329]: Won't do — this pattern is correct and NetBox-supported. NetBox overrides the model constructor specifically to accept these:

```python
# dcim/models/cables.py
def __init__(self, *args, a_terminations=None, b_terminations=None, **kwargs):
```

So `Cable(a_terminations=[...], b_terminations=[...])` is the intended NetBox API (used throughout NetBox's own test suite), not a non-field kwarg that `Model.__init__` would reject. Empirically it does not raise: this exact form runs in the pre-existing `setUpTestData` (`test_cable_matcher.py:217`) before every test in the class, and the full suite is **419 tests OK** — a `TypeError` at construction would error these tests, not pass them.

File: netbox_diode_plugin/api/matcher.py [hunk 13]
```diff
@@ -122,6 +122,7 @@ def invalidate_find_obj_entry(object_type: str, object_id: int):
 # replicas - that would require a DB unique constraint or a
 # coordinating lock.
 _REQUIRES_PRE_SAVE_MATCH = frozenset({
+    "dcim.cable",
```
[the remark this answers is not in the document]
Comment (leoparente) [reply netbox_diode_plugin/api/matcher.py:125]: Valid — fixed in 19e421b. Correct diagnosis: my `_align_cable_ends` only guards the plan-time (differ) path, but adding `dcim.cable` to `_REQUIRES_PRE_SAVE_MATCH` opened a second, apply-time path where a stale CREATE matches the existing cable and `_try_find_and_update_existing_instance()` re-saves the submitted terminations as-is, rewriting `cable_end`.

Fix: since the matcher finds the cable by its **A/B-insensitive termination set**, the set is identical by construction — so on the pre-save-match path I now strip `a_terminations`/`b_terminations` from the update (`_strip_matched_cable_terminations`). Only attributes update; the existing end assignment is left intact, so a stale end-swapped CREATE can't toggle `cable_end`. A genuine termination-set change wouldn't match this cable in the first place (different set → normal CREATE / clean conflict), so nothing legitimate is lost.

Regression test `test_stale_swapped_create_preserves_cable_ends`: applies a stale swapped CREATE for an existing cable and asserts `cable_end` is unchanged while a changed `label` still applies through this path (fails before the fix). Suite 421 OK.

**Comment by leoparente** [comment]
@codex review

File: netbox_diode_plugin/api/transformer.py [hunk 14]
```diff
@@ -199,14 +202,48 @@ def is_supported(field_name, ref_info):
...
+                    # camelCase protoJSON is normalized at the top level only;
+                    # the nested variant key needs its own normalization.
+                    variant_key = _camel_to_snake_case(raw_variant_key)
+                    concrete_type = get_generic_object_variant(variant_key)
+                    if concrete_type is None:
+                        node['_warnings'][field_name] = node['_warnings'].get(field_name, []) + [
+                            f"Skipping unknown generic-object variant key: {variant_key!r}"
+                        ]
+                        continue
+                    if concrete_type not in supported_models:
```
[the remark this answers is not in the document]
Comment (leoparente) [reply netbox_diode_plugin/api/transformer.py:238]: Fixed in 511f819. The GenericObject variant map spans all 98 content types, so a supported-but-non-terminable variant (e.g. `object_device` → `dcim.device`) was accepted and recursed before the cable serializer rejected it.

The cable generic-object arm now gates on `_cable_terminable_types()`, resolved from NetBox's own `dcim.constants.CABLE_TERMINATION_MODELS` (cached) so it tracks NetBox rather than a hand-maintained list — exactly the 9 certified endpoint types. Non-terminable variants now warn + skip that termination up front (before any child node is created), consistent with the existing unknown/unsupported-variant handling. Regression test added: `object_device` in `a_terminations` is skipped with no child device node created, while the valid interface termination is kept.

One correction on the orphan concern: `_apply_one_changeset` wraps the entire changeset (the child node **and** the cable) in a single `transaction.atomic()`, so a cable failure already rolls the child back rather than leaving it behind. But accepting non-terminable variants was still wrong — wasted create/rollback work and a confusing apply-time error instead of a clean skip — so rejecting up front as you suggested is the right fix. Suite 422 OK.

**Comment by leoparente** [comment]
@codex review

File: netbox_diode_plugin/api/matcher.py [hunk 15]
```diff
@@ -688,6 +695,120 @@ def build_queryset(self, data: dict) -> models.QuerySet:
...
+                pairs.append((content_type_id(object_type), object_id))
+
+        if not pairs:
+            return None
+
+        qs = self.model_class.objects.annotate(
+            _term_count=models.Count("terminations")
+        ).filter(_term_count=len(pairs))
+        for ct_id, pk in pairs:
+            qs = qs.filter(
```
[the remark this answers is not in the document]
Comment (leoparente) [reply netbox_diode_plugin/api/matcher.py:814]: Valid — fixed in da6c363. Confirmed the false match: with a duplicate pair, `len(pairs)` is inflated and each per-pair existence filter is a separate join, so both can be satisfied by the single `if1` row of an existing `if1<->if2` cable — the invalid `A:[if1] B:[if1]` payload matched that cable, and (with `dcim.cable` now in the pre-save-match path) the applier treated the invalid CREATE as an existing-cable update and stripped terminations instead of rejecting it.

A cable can't terminate the same object twice (the unique `(termination_type, termination_id)` constraint), so a duplicate pair describes an impossible cable. `build_queryset` now returns `None` when `len(set(pairs)) != len(pairs)` — not authoritatively matchable, so it falls through to CREATE and the serializer/`Cable.clean()` rejects the payload with a proper deviation.

Regression test asserts `A:[if1] B:[if1]` no longer matches the `if1<->if2` cable (returns a QuerySet containing that cable before the fix). Suite 423 OK.

**Comment by leoparente** [comment]
@codex review

File: netbox_diode_plugin/api/matcher.py [hunk 16]
```diff
@@ -688,6 +695,129 @@ def build_queryset(self, data: dict) -> models.QuerySet:
...
+        count, then require every requested (termination_type ct_id,
+        termination_id pk) is present. Rejects subset/superset. Returns None if
+        any object_id is still unresolved (cannot authoritatively match).
+        """
+        if not self.has_required_fields(data):
+            return None
+
+        pairs = []  # (content_type_id, pk)
+        for field in (self.a_field, self.b_field):
+            for term in data.get(field, []):
```
[the remark this answers is not in the document]
Comment (leoparente) [reply netbox_diode_plugin/api/matcher.py:789]: Valid — fixed in c5d08ff. Confirmed the real defect is a **cross-path inconsistency**: for the same repartition input on a ≥3-termination cable, the differ UPDATE path (matched by `netbox_id`) *applies* it (moves the termination A→B, serializer accepts the `cable_end` move), but the applier CREATE pre-save-match path *silently dropped* it — `build_queryset` flattens both ends, so the repartitioned payload matched the same cable, and `_strip_matched_cable_terminations` unconditionally dropped the terminations.

I kept the matcher's deliberate A/B-insensitive set identity untouched and made the **strip decision** grouping-aware instead: `_strip_matched_cable_terminations` now strips only when the submitted A/B grouping equals the existing one **up to a whole-end swap** (`_cable_partition_matches` — identical, within-end reorder, or pure swap), which preserves the no-toggle fix for those cases. A genuine repartition (same set, different grouping) is passed through so the serializer applies it — making this path consistent with the differ.

Regression tests added: repartition is applied (`if2` moves A→B); whole-end swap and within-end reorder stay no-ops. Suite 426 OK.

One scope note: this normalizes the A/B **grouping** (available today from `cable_end`). Positional **order within** a single end is still treated as no-change — that needs the per-termination connector/lane proto work tracked in **OBS-3411**.

**Comment by leoparente** [comment]
@codex review

**Review by jajeffries — APPROVED** [review]

**Review by mfiedorowicz — APPROVED** [review]
